vllm/v1/cudagraph_dispatcher.py

```diff
@@ -0,0 +1,144 @@
+# SPDX-License-Identifier: Apache-2.0
+# SPDX-FileCopyrightText: Copyright contributors to the vLLM project
+from typing import Any, Callable, NamedTuple, Optional
+
+from vllm.compilation.cuda_graph import CUDAGraphOptions, CUDAGraphWrapper
+from vllm.config import (CompilationLevel, CUDAGraphMode,
+                         CUDAGraphRuntimeStyle, VllmConfig)
+from vllm.logger import init_logger
+
+logger = init_logger(__name__)
+
+
+class DispatchKey(NamedTuple):
+    """
+    Key for dispatching cudagraphs.
+    """
+    cudagraph_runtime_style: CUDAGraphRuntimeStyle
+    # Be aware that is_pure_decode should be default None
+    # for both piecewise cudagraphs and no cudagraphs.
+    is_pure_decode: Optional[bool] = None
+
+
+class CudagraphDispatcher:
+    """
+    Runtime cudagraph dispatcher to switch between multiple cudagraphs.
+    """
+
+    def __init__(self, vllm_config: VllmConfig):
+        self.vllm_config = vllm_config
+        self.compilation_config = vllm_config.compilation_config
+        self.cudagraph_mode = self.compilation_config.cudagraph_mode
+
+        self.model: Callable = None  # type: ignore
+        # we lazy initialize self.model once the model loading of
+        # runner have been done.
+
+        # Dict to store cudagraph candidates for runtime dispatching.
+        self.cudagraph_candidates: dict[DispatchKey, Any] = {}
```

Comment:
nit: can we name DispatchKey something else? This means something very specific in PyTorch.